### PR TITLE
add fleet metric

### PR DESCRIPTION
**Pull request by xjqbest**
### PR types
New features

### PR changes
APIs

### Describe
add fleet metric，auc／acc／mse／mae／rmae／sum／max／min

### CODE REVIEW

File: python/paddle/fleet/metrics/metric.py [hunk 1]
```diff
@@ -11,3 +11,311 @@
...
+          # in model.py
+          input = fluid.layers.cast(some_input, dtype='float32')
+          cnt = fluid.layers.reduce_sum(input)
+          global_cnt = fluid.layers.create_global_var(persistable=True, dtype='float32', shape=[1], value=0)
+          tmp = fluid.layers.elementwise_add(cnt, global_cnt)
+          fluid.layers.assign(tmp, global_cnt)
+          
+          # in train.py, after train or infer
+          res = np.array(scope.find_var(global_cnt.name).get_tensor())
+          print("sum array: ", paddle.fleet.sum(res))
```
Comment (guru4elephant) [inline python/paddle/fleet/metrics/metric.py:46]: can you make `input` compatible with name string, variable, numpy array. and can have a default scope for variable finding

File: python/paddle/fleet/metrics/metric.py [hunk 2]
```diff
@@ -11,3 +11,311 @@
...
+          # in model.py
+          input = fluid.layers.cast(some_input, dtype='float32')
+          cnt = fluid.layers.reduce_sum(input)
+          global_cnt = fluid.layers.create_global_var(persistable=True, dtype='float32', shape=[1], value=0)
+          tmp = fluid.layers.elementwise_max(cnt, global_cnt)
+          fluid.layers.assign(tmp, global_cnt)
+
+          # in train.py, after train or infer
+          res = np.array(scope.find_var(global_cnt.name).get_tensor())
+          print("max array: ", paddle.fleet.max(res))
```
Comment (guru4elephant) [inline python/paddle/fleet/metrics/metric.py:85]: same as sum

File: python/paddle/fleet/metrics/metric.py [hunk 3]
```diff
@@ -11,3 +11,311 @@
...
+
+          # in train.py, after train or infer
+          res = np.array(scope.find_var(global_cnt.name).get_tensor())
+          print("min array: ", paddle.fleet.min(res))
+    """
+    fleet._role_maker._barrier_worker()
+    old_shape = np.array(input.shape)
+    output = np.copy(input) * 0
+    fleet._role_maker._all_reduce(input, output, mode="min")
+    output = output.reshape(old_shape)
```
Comment (guru4elephant) [inline python/paddle/fleet/metrics/metric.py:136]: same as sum

File: python/paddle/fleet/metrics/metric.py [hunk 4]
```diff
@@ -11,3 +11,311 @@
...
+          similarity_norm = fluid.layers.sigmoid(fluid.layers.clip(output, min=-15.0, max=15.0))
+          binary_predict = fluid.layers.concat(
+              input=[fluid.layers.elementwise_sub(fluid.layers.ceil(similarity_norm), similarity_norm), similarity_norm], axis=1)
+          self.auc, batch_auc, [batch_stat_pos, batch_stat_neg, stat_pos, stat_neg] =
+              fluid.layers.auc(input=binary_predict, label=label, curve='ROC', num_thresholds=4096)
+
+          # in train.py, after train or infer
+          pos = np.array(scope.find_var(stat_pos.name).get_tensor())
+          neg = np.array(scope.find_var(stat_neg.name).get_tensor())
+          print("auc: ", paddle.fleet.auc(pos, neg))
```
Comment (guru4elephant) [inline python/paddle/fleet/metrics/metric.py:165]: make this compatible with string/variable and numpy.

File: python/paddle/fleet/metrics/metric.py [hunk 5]
```diff
@@ -11,3 +11,311 @@
...
+
+    Example:
+        .. code-block:: python
+
+          # in model.py
+          sqrerr, abserr, prob, q, pos, total = fluid.contrib.layers.ctr_metric_bundle(similarity_norm, fluid.layers.cast(x=label, dtype='float32'))
+
+          # in train.py, after train or infer
+          res = np.array(scope.find_var(abserr.name).get_tensor())
+          print("mae: ", paddle.fleet.mae(res, total_ins_num))
```
Comment (guru4elephant) [inline python/paddle/fleet/metrics/metric.py:243]: the same here.

File: python/paddle/fleet/metrics/metric.py [hunk 6]
```diff
@@ -11,3 +11,311 @@
...
+
+    Example:
+        .. code-block:: python
+
+          # in model.py
+          sqrerr, abserr, prob, q, pos, total = fluid.contrib.layers.ctr_metric_bundle(similarity_norm, fluid.layers.cast(x=label, dtype='float32'))
+
+          # in train.py, after train or infer
+          res = np.array(scope.find_var(sqrerr.name).get_tensor())
+          print("rmse: ", paddle.fleet.rmse(res, total_ins_num))
```
Comment (guru4elephant) [inline python/paddle/fleet/metrics/metric.py:281]: same here.

File: python/paddle/fleet/metrics/metric.py [hunk 7]
```diff
@@ -11,3 +11,311 @@
...
+
+    Example:
+        .. code-block:: python
+
+          # in model.py
+          sqrerr, abserr, prob, q, pos, total = fluid.contrib.layers.ctr_metric_bundle(similarity_norm, fluid.layers.cast(x=label, dtype='float32'))
+
+          # in train.py, after train or infer
+          metric = np.array(scope.find_var(sqrerr.name).get_tensor())
+          print("mse: ", paddle.fleet.mse(metric, total_ins_num))
```
Comment (guru4elephant) [inline python/paddle/fleet/metrics/metric.py:319]: save here.

File: python/paddle/fleet/metrics/metric.py [hunk 8]
```diff
@@ -11,3 +11,311 @@
...
+          fluid.layers.assign(tmp1, global_correct)
+
+          global_total = fluid.layers.create_global_var(persistable=True, dtype='float32', shape=[1], value=0)
+          tmp2 = fluid.layers.elementwise_min(total, global_total)
+          fluid.layers.assign(tmp2, global_total)
+
+          # in train.py, after train or infer
+          correct_num = np.array(scope.find_var(correct.name).get_tensor())
+          total_num = np.array(scope.find_var(total.name).get_tensor())
+          print("accuracy: ", paddle.fleet.acc(correct_num, total_num))
```
Comment (guru4elephant) [inline python/paddle/fleet/metrics/metric.py:368]: same here.

**Review by guru4elephant — COMMENTED** [review]
please make the interface compatible with string, variable and numpy array to make the usage more easily.

**Review by guru4elephant — APPROVED** [review]
LGTM

**Comment by guru4elephant** [comment]
please add Chinese Documents

**Comment by xjqbest** [comment]
> please add Chinese Documents

ok，thanks

**Pull request merged**